iroh-pkarr-node-discovery/src/lib.rs

```diff
@@ -76,83 +93,245 @@ impl<'a> Builder<'a> {
         self
     }
 
+    /// Set the time-to-live value for the DNS packets.
+    pub fn ttl(mut self, ttl: u32) -> Self {
+        self.ttl = Some(ttl);
+        self
+    }
+
+    /// Set the pkarr relay URL to use.
+    pub fn pkarr_relay(mut self, pkarr_relay: Url) -> Self {
+        self.pkarr_relay = Some(pkarr_relay);
+        self
+    }
+
+    /// Use the default pkarr relay URL.
+    pub fn iroh_pkarr_relay(mut self) -> Self {
+        self.pkarr_relay = Some(N0_DNS_PKARR_RELAY.parse().expect("valid URL"));
+        self
+    }
+
+    /// Set whether to publish to the mainline DHT.
+    pub fn dht(mut self, dht: bool) -> Self {
+        self.dht = dht;
+        self
+    }
+
     /// Build the discovery mechanism.
-    pub fn build(self) -> PkarrNodeDiscovery {
+    pub fn build(self) -> anyhow::Result<PkarrNodeDiscovery> {
         let client = self.client.unwrap_or_default();
-        PkarrNodeDiscovery::new(client, self.secret_key)
+        let ttl = self.ttl.unwrap_or(DEFAULT_PKARR_TTL);
+        let pkarr_relay = self.pkarr_relay;
+        let dht = self.dht;
+        anyhow::ensure!(
+            dht || pkarr_relay.is_some(),
+            "at least one of DHT or relay must be enabled"
+        );
+        Ok(PkarrNodeDiscovery::new(
+            client,
+            self.secret_key,
+            ttl,
+            pkarr_relay,
+            dht,
+        ))
     }
 }
 
 impl PkarrNodeDiscovery {
-    pub fn builder() -> Builder<'static> {
+    /// Create a new builder for PkarrNodeDiscovery.
+    pub fn builder() -> Builder {
         Builder::default()
     }
 
     /// Create a new discovery mechanism.
     ///
     /// If a secret key is provided, the node will publish its address to the DHT.
     /// If no secret key is provided, publish will be a no-op, but resolving other nodes will still work.
-    fn new(pkarr: PkarrClient, secret_key: Option<&SecretKey>) -> Self {
-        let keypair =
-            secret_key.map(|secret_key| pkarr::Keypair::from_secret_key(&secret_key.to_bytes()));
+    fn new(
+        pkarr: PkarrClient,
+        secret_key: Option<SecretKey>,
+        ttl: u32,
+        relay_url: Option<Url>,
+        dht: bool,
+    ) -> Self {
         Self(Arc::new(Inner {
-            keypair,
+            secret_key,
             pkarr,
+            ttl,
+            relay_url,
+            dht,
             task: Default::default(),
         }))
     }
+
+    /// Periodically publish the node address to the DHT and relay.
+    async fn publish_loop(self, keypair: SecretKey, signed_packet: SignedPacket) {
+        let this = self;
+        let z32 = pkarr::PublicKey::try_from(*keypair.public().as_bytes())
+            .expect("valid public key")
+            .to_z32();
+        loop {
+            // initial delay. If the task gets aborted before this delay is over,
+            // we have not published anything to the DHT yet.
+            tokio::time::sleep(INITIAL_PUBLISH_DELAY).await;
```

Comment:
why not just put this outside of the loop and then skip the subtraction in the second sleep?

Reply:
yeah, you're right.